pkg/volume/util/operationexecutor/operation_executor.go

```diff
@@ -0,0 +1,763 @@
+			err)
+	}
+
+	return func() error {
+		// Execute detach
+		detachErr := volumeDetacher.Detach(deviceName, volumeToDetach.NodeName)
+
+		if detachErr != nil {
+			// On failure, just log and exit. The controller will retry
+			glog.Errorf(
+				"DetachVolume.Detach failed for volume %q (spec.Name: %q) from node %q with: %v",
+				volumeToDetach.VolumeName,
+				volumeToDetach.VolumeSpec.Name(),
+				volumeToDetach.NodeName,
+				detachErr)
+			return detachErr
+		}
+
+		glog.Infof(
+			"DetachVolume.Detach succeeded for volume %q (spec.Name: %q) from node %q.",
+			volumeToDetach.VolumeName,
+			volumeToDetach.VolumeSpec.Name(),
+			volumeToDetach.NodeName)
+
+		// Update actual state of world
+		actualStateOfWorld.MarkVolumeAsDetached(
+			volumeToDetach.VolumeName, volumeToDetach.NodeName)
+
+		return nil
+	}, nil
+}
+
+func (oe *operationExecutor) generateMountVolumeFunc(
+	waitForAttachTimeout time.Duration,
+	volumeToMount VolumeToMount,
+	actualStateOfWorld ActualStateOfWorldMounterUpdater) (func() error, error) {
+	// Get mounter plugin
+	mountableVolumePlugin, err :=
+		oe.volumePluginMgr.FindPluginBySpec(volumeToMount.VolumeSpec)
+	if err != nil || mountableVolumePlugin == nil {
+		return nil, fmt.Errorf(
+			"MountVolume.FindPluginBySpec failed for volume %q (spec.Name: %q) pod %q (UID: %q) with: %v",
+			volumeToMount.VolumeName,
+			volumeToMount.VolumeSpec.Name(),
+			volumeToMount.PodName,
+			volumeToMount.Pod.UID,
+			err)
+	}
+
+	volumeMounter, newMounterErr := mountableVolumePlugin.NewMounter(
+		volumeToMount.VolumeSpec,
+		volumeToMount.Pod,
+		volume.VolumeOptions{})
+	if newMounterErr != nil {
+		return nil, fmt.Errorf(
+			"MountVolume.NewMounter failed for volume %q (spec.Name: %q) pod %q (UID: %q) with: %v",
+			volumeToMount.VolumeName,
+			volumeToMount.VolumeSpec.Name(),
+			volumeToMount.PodName,
+			volumeToMount.Pod.UID,
+			newMounterErr)
+	}
+
+	// Get attacher, if possible
+	attachableVolumePlugin, _ :=
+		oe.volumePluginMgr.FindAttachablePluginBySpec(volumeToMount.VolumeSpec)
+	var volumeAttacher volume.Attacher
+	if attachableVolumePlugin != nil {
+		volumeAttacher, _ = attachableVolumePlugin.NewAttacher()
+	}
+
+	var fsGroup *int64
+	if volumeToMount.Pod.Spec.SecurityContext != nil &&
+		volumeToMount.Pod.Spec.SecurityContext.FSGroup != nil {
+		fsGroup = volumeToMount.Pod.Spec.SecurityContext.FSGroup
+	}
+
+	return func() error {
+		if volumeAttacher != nil {
+			// Wait for attachable volumes to finish attaching
+			glog.Infof(
+				"Entering MountVolume.WaitForAttach for volume %q (spec.Name: %q) pod %q (UID: %q).",
+				volumeToMount.VolumeName,
+				volumeToMount.VolumeSpec.Name(),
+				volumeToMount.PodName,
+				volumeToMount.Pod.UID)
+
+			devicePath, err := volumeAttacher.WaitForAttach(
+				volumeToMount.VolumeSpec, waitForAttachTimeout)
+			if err != nil {
+				glog.Errorf(
+					"MountVolume.WaitForAttach failed for volume %q (spec.Name: %q) pod %q (UID: %q) with: %v",
```

Comment:
`MountVolume` is the wrapper call. It results in `WaitForAttach()`, `MountDevice()`, and `Setup()`